### ISSUE TITLE
Browser crashes on accessing big files

### ISSUE THREAD

**Issue by MarionMoseby** [go-gitea/gitea#12141]
- Gitea version (or commit ref): Codeberg?
- Git version: Codeberg?
- Operating system: Ubuntu 20.04
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (Warning: The following link might crash your browser: [LINK](https://try.gitea.io/FlyingFlamingo/BioPython-Analyzer/src/branch/master/Sample%20Input/Subject/genomic%20%282%29.gbff))
  - [ ] No
  - [ ] Not relevant
- Log gist: Im sorry I dont know what this is

## Description

[I've previously reported this bug on codeberg](
https://codeberg.org/Codeberg/Community/issues/201), a gitea instance, and I have been redirected to this issue tracker to report it here as well. The thing is, When I open this file for example, firefox becomes unresponsive due to the file being too big. Maybe the maximum file size to be shown in the preview should be decreased? In GH at least I have no option to preview such a large file.

## Screenshots

I cannot add an screenshot as essentially my computer crashes, but basically the tab turns white and all slows down

**Comment by lafriks** [comment]
How big the file is?

**Comment by silverwind** [comment]
Big files are certainly an issue. GitHub has a limit of 1MB for file display above which it offers only a link to the raw file.

**Comment by mrsdizzie** [comment]
GitHub limit For display is 5mb and stop highlighting at 1MB

https://help.github.com/en/github/creating-cloning-and-archiving-repositories/limits-for-viewing-content-and-diffs-in-a-repository

Gitea def has some other bug with larger files I've seen when working on syntax highlighting but not time to look into yet. loading a few mb of non highlighted text file should be fine and is no issue on github or other sites. 

Might be something JavaScript side as I get one of those "this page is using too much memory" warnings sometimes. Will try and see if I can reproduce and find cause.

**Comment by silverwind** [comment]
> GitHub limit For display is 5mb and stop highlighting at 1MB

Not true or outdated info. See these two examples. One is slightly below 1MB, one is above:

https://github.com/StylishThemes/GitHub-Dark/blob/master/github-dark.css
https://github.com/StylishThemes/GitHub-Dark/blob/master/github-dark.user.css

**Comment by mrsdizzie** [comment]
@silverwind interesting yea, I guess they have changed that. Not sure why they say "right now" if that is always the limit.

If that is the case, it seems like they just have a single limit of 1MB and there isn't a situation where they show file content but don't highlight it. We could have that by just changing this default to 1MB instead of the current 8:

https://github.com/go-gitea/gitea/blob/c5b08f6d5a73e6ba84da84e804d05a8dd3d651be/modules/setting/setting.go#L214

Which would then match the already 1MB limit for highlighting.

Looking at the actual cause of slowness for large files in Gitea it is all 'recalculating styles' in the browser -- so there is something which doesn't scale there (and maybe thats fine, probably why Github has a 1MB limit too then).

**Comment by silverwind** [comment]
GitHub also has a limit at 20000 line number after which they just stop highlighting. See [this example](https://github.com/StylishThemes/GitHub-Dark/commit/2e78515b306c1d74287516d70f571eaab8acbf14). You'll notice that highlighting stops at line 20k. It's also visible in simple diffs like [this one](https://github.com/StylishThemes/GitHub-Dark/commit/9c6d10c67c7cbc552108a513ecefb353ada88c6a) when changes happen after line 20k. I'm not sure if we'll need a similar limit with our approach.

> 'recalculating styles'

Maybe some JS interference. Probably worth a try viewing diffs with scripts disabled. There's also [this nasty bit of JS](https://github.com/go-gitea/gitea/blob/master/templates/repo/diff/box.tmpl#L277) we should get rid of for split diff.

**Comment by MarionMoseby** [comment]
Any progress with these? I feel like this is a valid issue and should not be closed

**Comment by schorsch13** [comment]
Another codeberg user reported a related issue https://codeberg.org/Codeberg/Community/issues/531 where the browser crashes when accessing [this link](https://codeberg.org/boud/gevolution/compare/0906a1e...fe756e7#diff-6) WARNING: your browser might crash! But this link directs you to a git-diff instead of a big file

**Comment by silverwind** [comment]
Yes, we probably need some kind of lazy-loading solution. The HTML renders pretty fast in this case at least:

```bash
$ time curl -s 'https://codeberg.org/boud/gevolution/compare/0906a1e...fe756e7#diff-6' > /dev/null
curl -s  > /dev/null  0.03s user 0.04s system 3% cpu 2.200 total
```

**Comment by zeripath** [comment]
We already have a lazy load system on 1.16

Reduce the MAX_GIT_DIFF_FILES and MAX_GIT_DIFF_LINES as appropriate and use the functionality in 1.16 to load large diffs as needed.

Iirc in 1.16 I also made split diffs a whole lot less burdensome to render.
